### PR TITLE
[flang][OpenMP] Handle expressions in `target ... do` loop control

### PR DESCRIPTION
**Pull request by ergawy**
When emitting the ops required to compute the target loop's trip count, flang might emit ops outside the target regions that operands defined inside the region. This is fixed up by `HostClausesInsertionGuard` already.

However, the current support only handles simple cases. If the loop control contains more elaborate expressions, the fix up logic does not handle it properly. This PR handles such cases.

### CODE REVIEW

**Review by agozillon — APPROVED** [review]
LGTM, I believe @skatrak covered everything with his comments! So please address them and wait on his approval prior to merging. 

I have one likely stupid question though, as I am a little naive with this segment of work, would this create additional unused omp.map.info operations as we previously had the defining operation inside of the target region and then lifted it out? I would imagine not, and the inverse is more likely. But I thought it was worth clarifying my understanding of the interaction as we have a segment of code that will clone the use chain into the target region if it can, but if it can't it'll generate map operations. However, this is sort of the inverse situation in a way. 

TL;DR: there's two ways we generate map.info operations implicitly, I am wondering if we end up generating map.info in this situation that we no longer require after this modification is performed!

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 1]
```diff
@@ -123,9 +124,85 @@ class HostClausesInsertionGuard {
...
   mlir::omp::TargetOp targetOp;
 
+  // Finds the list of op operands that escape the target op's region; that is:
+  // the operands that are used outside the target op but defined inside it.
+  void
+  findEscapingOpOperands(llvm::DenseSet<mlir::OpOperand *> &escapingOperands) {
+    if (!targetOp)
+      return;
+
+    mlir::Block *targetParentBlock = targetOp->getBlock();
```
[the remark this answers is not in the document]
Comment (ergawy) [reply flang/lib/Lower/OpenMP/OpenMP.cpp]: Done.

**Comment by ergawy** [comment]
> ... would this create additional unused omp.map.info operations as we previously had the defining operation inside of the target region and then lifted it out? ...

Not a stupid question at all 😛!

We do not create any additional `omp.map.info` ops since we do not actually **lift** the defining op but **clone it** (along with the slice of ops that define its operands) and then after cloning we remap the operands of the cloned slice to use values defined outside the target region.

**Comment by agozillon** [comment]
> > ... would this create additional unused omp.map.info operations as we previously had the defining operation inside of the target region and then lifted it out? ...
> 
> Not a stupid question at all 😛!
> 
> We do not create any additional `omp.map.info` ops since we do not actually **lift** the defining op but **clone it** (along with the slice of ops that define its operands) and then after cloning we remap the operands of the cloned slice to use values defined outside the target region.

Thank you for explaining that and clarifying it for me!

**Review by skatrak — APPROVED** [review]
Thank you Kareem for your clarifications, LGTM!
